src/rank_llm/retrieve/retriever.py

```diff
@@ -84,15 +91,38 @@ def from_saved_results(file_name: str):
             )
         retriever = Retriever(RetrievalMode.SAVED_FILE, dataset=retrieved_results)
         return retriever.retrieve()
+    
+    @staticmethod
+    def from_custom_index(index_path: str, topics_path: str, index_type: str):
+        if not index_path:
+            raise ValueError("Please provide a path to the index")
+        if not topics_path:
+            raise ValueError("Please provide a path to the topics file")
+        if index_type not in ['lucene', 'impact']:
+            raise ValueError(f"index_type must be [lucene, impact], not {index_type}")
+        
+        retriever = Retriever(
+            RetrievalMode.CUSTOM, 
+            dataset=None, 
+            retrieval_method=RetrievalMethod.UNSPECIFIED, 
+            index_path=index_path, 
+            topics_path=topics_path, 
+            index_type=index_type
+        )
+        return retriever.retrieve()
 
     def retrieve(self) -> List[Dict[str, Any]]:
-        if self._retrieval_mode == RetrievalMode.DATASET:
+        if self._retrieval_mode == RetrievalMode.DATASET or self._retrieval_mode == RetrievalMode.CUSTOM:

```

Comment:
you should not merge these two since the dataset mode uses retrieval method and dataset name for saving the results, this will fail for the custom mode.
For custom index path, use a candidate filename based on the provided input parameters.

Reply:
good catch, I refactored it and now we use the directory name of the index as the 'dataset' for the candidate filename